### ISSUE TITLE
IDA Crashes when syncing structures

### ISSUE THREAD

**Issue by sday15** [IDArlingTeam/IDArling#78]
2 uses are using IDArling to work on a database.

If user A creates or edits any structure (including renaming fields) then the IDA instance for user B will crash.

From turning the server logging level up to DEBUG it looks like the IDA instance crashes either during or after receiving the DefaultEvent packet that contains the full information of all the structures in user A's database.

Here is a redacted extract from the server log:

```
[DEBUG] (10.45.3.140:41910) Received packet: DefaultEvent(sname=struc_abc, event_type=struc_member_renamed, offset=0, newname=testing, type=event)
[DEBUG] (10.45.3.123:43426) Sending packet: DefaultEvent(sname=struc_abc, event_type=struc_member_renamed, offset=0, newname=testing, type=event)
[DEBUG] (10.45.3.140:41910) Received packet: DefaultEvent(local_types=[[1, '__n64', '\x1dQÑ\x1b\x02%\x1b\x03$\x1b\x05#\x1b\t"\x1b
<CUT>
[DEBUG] (10.45.3.123:43426) Sending packet: DefaultEvent(local_types=[[1, '__n64', '\x1dQÑ\x1b\x02%\x1b\x03$\x1b\x05#\x1b\t"\x1b
<CUT>
```

This issue appears to be 100% repeatable across all databases. We are using IDA 7.1 on Ubuntu with the standalone server.

**Comment by sday15** [comment]
Update: This only appears to happen one way when user A creates or edits structures in their database. If user B makes the change then this is correctly synced to user A's database and no crash occurs.

**Comment by patateqbool** [comment]
Hi sday15,
Thanks for the bug report, I'll push a patch during the day, let me know if it's fix your problem please

**Comment by patateqbool** [comment]
Hi @sday15,
I've tried to fix the bug, unfortunately LocalTypes and user-defined Struct/Enum conflict...
I'm still working on a solution for this issue.

**Comment by NyaMisty** [comment]
We can try to unhook when one of my them occurred?

patateqbool <notifications@github.com>于2018年11月19日 周一15:48写道：

**Comment by patateqbool** [comment]
Sorry, I'm not sure to understand the sentence correctly, but here is my answer:

When creating a struct/enum in struc/enum view, first a LocalTypesChanged event is sent, then a Struct/enum related event is sent. For the moment, my only idea is to "queue" the LocalTypesChanged event, wait for the next event, if the next event is a Struct/enum related event, drop the LocalTypesChanged event, else send LocalTypesChanged event then the next event.

**Comment by NyaMisty** [comment]
Yeah, as the Local Types part was first implemented by me, I noticed this
before, and it’s in fact the source of lots of struct issues. I mean that
we can unhook the Struct related hooks and re-enable it afterwards.
Or we can consider simply remove the struct hooks as it’s connected with
local types internally?

patateqbool <notifications@github.com>于2018年11月20日 周二22:01写道：

**Comment by patateqbool** [comment]
I agree, I'm going to push a fix that disables localTypes tonight until we have a stable fix to integrate them.

**Comment by sday15** [comment]
disabling localTypes appears to have fixed the crashing issue.

**Comment by patateqbool** [comment]
Good to know, I am always looking for a solution to properly integrate localTypes.
Thanks for the feedback :)
